### PR TITLE
[Master] Add bbe for RegExp type

### PR DESCRIPTION
**Pull request by LakshanWeerasinghe**
## Purpose
$subject

Update 5 PR #4193

### CODE REVIEW

File: examples/regexp-type/regexp.bal [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+public function main() returns error? {
+    
+    // Declare a RegExp using string module
+    string:RegExp _ = re `[a-zA-Z0-9]`;
+
+    // Declare a RegExp using `fromString` method in `lang.regexp` module
+    string:RegExp pattern = check regexp:fromString("HELLO*");
+    io:println("Pattern `HELLO*` matches HELLO ", pattern.isFullMatch("HELLO"));
+
+    boolean result1 = re `a+`.isFullMatch("aaa");
```
Comment (SasinduDilshara) [inline examples/regexp-type/regexp.bal:13]:
```suggestion
    boolean result = re `a+`.isFullMatch("aaa");
```

File: examples/regexp-type/regexp.bal [hunk 2]
```diff
@@ -0,0 +1,42 @@
+import ballerina/io;
+import ballerina/lang.regexp;
+
+public function main() returns error? {
+    
+    // Declare a RegExp using string module
```
Comment (SasinduDilshara) [inline examples/regexp-type/regexp.bal:6]:
```suggestion
    // Declare a RegExp using string module.
```

File: examples/regexp-type/regexp.metatags [hunk 3]
```diff
@@ -0,0 +1,2 @@
+description: This BBE demonstrates how to define a regular expression and different patterns supported by ballerina RegExp type.
+keywords: ballerina, ballerina by example, bbe, regex, regexp, regular expression
```
Comment (SasinduDilshara) [inline examples/regexp-type/regexp.metatags:2]: Need a new line here.

**Review by SasinduDilshara — CHANGES_REQUESTED** [review]

File: examples/regexp-type/regexp.bal [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    // Ballerina RegExp supports to write different unicode languages using unicode script patterns
+    // `\p{sc=...}` pattern is used
+    string:RegExp pattern5 = re `\p{sc=Latin}`;
+    regexp:Span? latinValue = pattern5.find("aεЛ");
+    io:println("Pattern `\\p{sc=Latin}` matches aεЛ ", latinValue is regexp:Span);
+
+    // Ballerina RegExp supports non-capturing groups to control the behavior of regular expression patterns
+    // `i` flag will ignore the case of the pattern
+    string:RegExp pattern6 = re `(?i:BalleRINA)`;
+    io:println("Pattern `(?i:BalleRINA)` matches xyz ", pattern6.isFullMatch("Ballerina"));
```
Comment (SasinduDilshara) [inline examples/regexp-type/regexp.bal:41]:
```suggestion
    io:println("Pattern `(?i:BalleRINA)` matches Ballerina ", pattern6.isFullMatch("Ballerina"));
```

**Review by SasinduDilshara — CHANGES_REQUESTED** [review]

**Comment by LakshanWeerasinghe** [comment]
@SasinduDilshara please refer #4193 for updated changes

**Comment by SasinduDilshara** [comment]
> @SasinduDilshara please refer #4193 for updated changes

Sure, Please cherrypick the relevant commits to here as well

File: examples/regexp-type/regexp_type.bal [hunk 5]
```diff
@@ -0,0 +1,42 @@
+import ballerina/io;
+import ballerina/lang.regexp;
+
+public function main() returns error? {
+    
+    // Declare a RegExp using the string module.
```
Comment (pcnfernando) [inline examples/regexp-type/regexp_type.bal:6]:
```suggestion
    // Declare a RegExp using regular expression literal.
    // string module contains a type alias for Regexp type defined in regexp module
```

File: examples/regexp-type/regexp_type.bal [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    string:RegExp _ = re `[a-zA-Z0-9]`;
+
+    // Declare a RegExp using the `fromString` method in the `lang.regexp` module.
```
Comment (pcnfernando) [inline examples/regexp-type/regexp_type.bal:9]:
```suggestion
    // Declare a RegExp using the `fromString` constructor method.
```

File: examples/regexp-type/regexp_type.bal [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    string:RegExp pattern = check regexp:fromString("HELLO*");
+    io:println("Pattern `HELLO*` matches HELLO ", pattern.isFullMatch("HELLO"));
+
+    boolean result = re `a+`.isFullMatch("aaa");
```
Comment (pcnfernando) [inline examples/regexp-type/regexp_type.bal:13]:
```suggestion
    // Matches any string that contains one or more consecutive "a" characters.
    boolean result = re `a+`.isFullMatch("aaa");
```

File: examples/regexp-type/regexp_type.bal [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+    io:println("Pattern `a+` matches aaa ", result);
+
+    // Interpolation can be used to change the pattern dynamically.
+    string literal = "xyz";
+    string:RegExp pattern2 = re `abc|${literal}`;
```
Comment (pcnfernando) [inline examples/regexp-type/regexp_type.bal:18]:
```suggestion
    // Matches any string that starts with "abc" or "XYZ".
    string:RegExp pattern2 = re `abc|${literal}`;
```

File: examples/regexp-type/regexp_type.bal [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+    io:println("Pattern `abc|${\"xyz\"}` matches xyz ", pattern2.isFullMatch("xyz"));
```
Comment (pcnfernando) [inline examples/regexp-type/regexp_type.bal:19]:
```suggestion
    // Matches any string that starts with "abc" or "XYZ".
    io:println("Pattern `abc|${\"xyz\"}` matches xyz ", pattern2.isFullMatch("xyz"));
```

File: examples/regexp-type/regexp_type.bal [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+
+    // Ballerina RegExp supports writing Unicode general category patterns.
```
Comment (pcnfernando) [inline examples/regexp-type/regexp_type.bal:21]:
```suggestion
    // RegExp supports writing Unicode general category patterns. 
```

File: examples/regexp-type/regexp_type.bal [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+    // Characters are matched based on their Unicode properties.
+    // The `\p{...}` pattern is used.
```
Comment (pcnfernando) [inline examples/regexp-type/regexp_type.bal:23]:
```suggestion
```

File: examples/regexp-type/regexp_type.bal [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+    string:RegExp pattern3 = re `\p{Ll}`;
+    io:println("Pattern `\\p{Ll}` matches a ", pattern3.isFullMatch("a"));
```
Comment (pcnfernando) [inline examples/regexp-type/regexp_type.bal:25]:
```suggestion
    io:println("Pattern `\\p{Ll}` matches `a`: ", "a".matches(pattern3));
```

File: examples/regexp-type/regexp_type.out [hunk 13]
```diff
@@ -0,0 +1,8 @@
+$ bal run regexp_type.bal
+Pattern `HELLO*` matches HELLO true
+Pattern `a+` matches aaa true
+Pattern `abc|${"xyz"}` matches xyz true
+Pattern `\p{Ll}` matches a true
```
Comment (pcnfernando) [inline examples/regexp-type/regexp_type.out:5]:
```suggestion
Pattern `\p{Ll}` matches `a`: true
```

File: examples/regexp-type/regexp_type.bal [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+
+    // The `\P{...}` escape sequence is used to get the negation of the pattern.
+    // This pattern will match every non-puncutation character.
+    string:RegExp pattern4 = re `\P{P}`;
+    io:println("Pattern `\\p{P}` matches 0 ", pattern4.isFullMatch("0"));
```
Comment (pcnfernando) [inline examples/regexp-type/regexp_type.bal:30]: Update as https://github.com/ballerina-platform/ballerina-distribution/pull/4190/files#r1143139666

File: examples/regexp-type/regexp_type.bal [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+
+    // Ballerina RegExp supports writing different Unicode languages using Unicode script patterns.
+    // The `\p{sc=...}` pattern is used.
```
Comment (pcnfernando) [inline examples/regexp-type/regexp_type.bal:33]:
```suggestion
```

File: examples/regexp-type/regexp_type.bal [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+    // The `\p{...}` pattern is used.
+    string:RegExp pattern3 = re `\p{Ll}`;
+    io:println("Pattern `\\p{Ll}` matches a ", pattern3.isFullMatch("a"));
+
+    // The `\P{...}` escape sequence is used to get the negation of the pattern.
+    // This pattern will match every non-puncutation character.
+    string:RegExp pattern4 = re `\P{P}`;
+    io:println("Pattern `\\p{P}` matches 0 ", pattern4.isFullMatch("0"));
+
+    // Ballerina RegExp supports writing different Unicode languages using Unicode script patterns.
```
Comment (pcnfernando) [inline examples/regexp-type/regexp_type.bal:32]:
```suggestion
    // RegExp supports writing different Unicode languages using Unicode script patterns.
```

File: examples/regexp-type/regexp_type.bal [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+    // The `\p{sc=...}` pattern is used.
+    string:RegExp pattern5 = re `\p{sc=Latin}`;
+    regexp:Span? latinValue = pattern5.find("aεЛ");
+    io:println("Pattern `\\p{sc=Latin}` matches aεЛ ", latinValue is regexp:Span);
```
Comment (pcnfernando) [inline examples/regexp-type/regexp_type.bal:36]: Print the actual substring from the Span

File: examples/regexp-type/regexp_type.bal [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+
+    // Ballerina RegExp supports non-capturing groups to control the behavior of regular expression patterns.
+    // The `i` flag will ignore the case of the pattern.
+    string:RegExp pattern6 = re `(?i:BalleRINA)`;
+    io:println("Pattern `(?i:BalleRINA)` matches Ballerina ", pattern6.isFullMatch("Ballerina"));
```
Comment (pcnfernando) [inline examples/regexp-type/regexp_type.bal:41]: use string:matches() instead

File: examples/regexp-type/regexp_type.md [hunk 19]
```diff
@@ -0,0 +1,16 @@
+# RegExp 
+
+``RegExp`` is a built-in subtype of ``anydata``.
+
+The `lang.regexp` module contains various functions related to regular expressions. You can learn more about these functions and how to use them by referring to the `lang.regexp` API documentation.
+
+You can declare `RegExp` variables in two ways either by using the `string` as the module prefix or by using the `regexp` langlib as the module prefix. In the second method, the `regexp` module needs to be imported explicitly.
```
Comment (pcnfernando) [inline examples/regexp-type/regexp_type.md:7]: Defining part should come first. Then the using part
```suggestion
`RegExp` is a built-in type in Ballerina that represents a regular expression. Regular expressions are commonly used for tasks such as validating input, searching and replacing text, and parsing data.

A `RegExp` value can be created by using the regexp template expression or calling the `fromString` method of the [lang.regexp](https://lib.ballerina.io/ballerina/lang.regexp/latest#fromString) module. 
```

File: examples/regexp-type/regexp_type.out [hunk 20]
```diff
@@ -0,0 +1,8 @@
+$ bal run regexp_type.bal
```
Comment (pcnfernando) [inline examples/regexp-type/regexp_type.out:1]: Check and update the outputs once the suggestions are addressed

**Review by pcnfernando — APPROVED** [review]

**Review by praneesha — APPROVED** [review]

**Review by SasinduDilshara — APPROVED** [review]

**Pull request merged**